### ISSUE TITLE
Auth >= 4.5 / DnsDist 1.7 : LUA View resolution changed?

### ISSUE THREAD

**Issue by carboneater** [PowerDNS/pdns#11215]
- Program: Authoritative + DNSDist
 - Issue type: Bug report

### Short description
We upgraded our Authoritative servers from 4.2 to 4.5 and 4.6-rc1
(DNSDist were also upgraded from 1.5 to 1.7)
Upgrade rationale was to make use of the newly exposed prometheus metrics from all our services
We're making heavy use of the LUA view() function to alter our replies based on query's origin
All our LUA records using the view function changed behavior from matching the correct subnet to resolving a private address even from public request.
example LUA record:
`A "view({{{'127.0.0.1'},{'172.19.32.5'}},{{'172.16.0.0/12'},{'172.19.32.5'}},{{'0.0.0.0/0'},{'52.154.61.152'}}})"`

### Environment

 - Operating system: Ubuntu 18.04.6 LTS
 - Software version: Authoritative > 4.5, DNSDist 1.7
 - Software source: PowerDNS repository

### Steps to reproduce

1. In Authoritative, create a LUA record reading `auc-web-2.fgf.cloud LUA A "view({{{'127.0.0.1'},{'172.19.32.5'}},{{'172.16.0.0/12'},{'172.19.32.5'}},{{'0.0.0.0/0'},{'52.154.61.152'}}})"`  
(This record, along with many others were working on 4.2)
2. Run a dnsdist using the mostly off-the-docs config of
```
setLocal('172.18.16.6:53')
setACL({'0.0.0.0/0', '::/0'}) -- Allow all IPs access

newServer({address="172.18.16.4:55", name="ace-dns-1", pool="auth", useClientSubnet=true})
newServer({address="172.18.16.5:55", name="ace-dns-2", pool="auth", useClientSubnet=true})
newServer({address="172.20.16.4:55", name="acc-dns-1", pool="auth", useClientSubnet=true})
newServer({address="172.20.16.5:55", name="acc-dns-2", pool="auth", useClientSubnet=true})
newServer({address="127.0.0.1:5300", name="recursor", pool="recursor", useClientSubnet=true})
setServerPolicy(leastOutstanding)

recursive_ips = newNMG()
recursive_ips:addMask('69.70.183.246')

addAction(NetmaskGroupRule(recursive_ips), PoolAction('recursor'))
addAction(AllRule(), PoolAction('auth'))

webserver('172.18.16.6:8083')
setWebserverConfig({ acl='172.16.0.0/12', statsRequireAuthentication=false })
```
3. dig @8.8.8.8 +short auc-web-1.fgf.cloud
> 172.19.32.4
4. panic

### Expected behaviour
Expected behaviour was that the LUA view() parsing stayed the same from 4.2 onwards, as there were no mentions of changes to those in the changelog
(Public resolved address should be 52.154.61.152 for the example LUA script provided)

### Actual behaviour

- dig @8.8.8.8 +short auc-web-1.fgf.cloud
> 172.19.32.4

- tcpdump between dnsdist & authoritative:
```
10:31:07.402473 IP 172.18.16.6.55674 > 172.18.16.4.55: UDP, length 36
10:31:07.402639 IP 172.18.16.4.55 > 172.18.16.6.55674: UDP, length 36
10:31:07.809008 IP 172.20.16.6.47817 > 172.18.16.4.55: UDP, length 36
10:31:07.809280 IP 172.18.16.4.55 > 172.20.16.6.47817: UDP, length 36
10:31:08.418235 IP 172.18.16.6.34211 > 172.18.16.4.55: UDP, length 36
10:31:08.418367 IP 172.18.16.4.55 > 172.18.16.6.34211: UDP, length 36
10:31:08.822708 IP 172.20.16.6.48368 > 172.18.16.4.55: UDP, length 36
10:31:08.822912 IP 172.18.16.4.55 > 172.20.16.6.48368: UDP, length 36
10:31:09.432584 IP 172.18.16.6.55303 > 172.18.16.4.55: UDP, length 36
10:31:09.432717 IP 172.18.16.4.55 > 172.18.16.6.55303: UDP, length 36
10:31:09.836316 IP 172.20.16.6.60345 > 172.18.16.4.55: UDP, length 36
10:31:09.836530 IP 172.18.16.4.55 > 172.20.16.6.60345: UDP, length 36
10:31:10.448186 IP 172.18.16.6.39123 > 172.18.16.4.55: UDP, length 36
10:31:10.448555 IP 172.18.16.4.55 > 172.18.16.6.39123: UDP, length 36
10:31:10.849968 IP 172.20.16.6.33566 > 172.18.16.4.55: UDP, length 36
10:31:10.850184 IP 172.18.16.4.55 > 172.20.16.6.33566: UDP, length 36
10:31:11.464225 IP 172.18.16.6.54137 > 172.18.16.4.55: UDP, length 36
10:31:11.464389 IP 172.18.16.4.55 > 172.18.16.6.54137: UDP, length 36
10:31:11.552247 IP 172.18.16.6.36059 > 172.18.16.4.55: UDP, length 53
10:31:11.553028 IP 172.18.16.4.55 > 172.18.16.6.36059: UDP, length 89
10:31:11.862279 IP 172.20.16.6.56941 > 172.18.16.4.55: UDP, length 36
10:31:11.862953 IP 172.18.16.4.55 > 172.20.16.6.56941: UDP, length 36
10:31:12.479448 IP 172.18.16.6.50859 > 172.18.16.4.55: UDP, length 36
10:31:12.479670 IP 172.18.16.4.55 > 172.18.16.6.50859: UDP, length 36
10:31:12.875841 IP 172.20.16.6.39072 > 172.18.16.4.55: UDP, length 36
10:31:12.875982 IP 172.18.16.4.55 > 172.20.16.6.39072: UDP, length 36
10:31:13.494424 IP 172.18.16.6.50077 > 172.18.16.4.55: UDP, length 36
10:31:13.494590 IP 172.18.16.4.55 > 172.18.16.6.50077: UDP, length 36
```
I'm assuming the use of EDNS Client Subnet changes something in the capture, hence it not being printed.

### Other information

I'm quite out of my depth here.
Should you need anything else, I'll gladly provide it.

**Comment by Habbie** [comment]
Are you sure the problem appears with both 4.5 and 4.6? Not only with 4.6?

**Comment by carboneater** [comment]
Hi Habbie!

> Does the problem appear with both 4.5 and 4.6?
After trying, and failing, to setup a test environment, I gave up and ran some tests on production.

I created a new LUA record
`localtest.fgf.cloud LUA A "view({{{'127.0.0.1'}, {'127.0.0.1'}}, {{'172.17.129.0/24'},{'172.17.129.5'}}, {{'172.20.225.0/24'},{'172.20.225.4'}}, {{'0.0.0.0/0'}, {'1.1.1.1'}}})"`

When I try to resolve it by calling the Authoritative server from various machines, I get the expected response.

That behavior works as expected on both 4.5.2 and 4.6-rc1

---

Further investigations showed I had issues when the queries were forwarded by `dnsdist`.

Lots of googling and grepping in the configuration files showed we reverted the `edns-subnet-processing` setting in authoritative.

Re-activating it solved my issues immediately.
... Shame on me.

Sorry for the disruption and keep up the good work!
